File: src/main/java/me/ghonix/practice/MedianTwoSortedArrays.java
package me.ghonix.practice;

/**
 * https://leetcode.com/problems/median-of-two-sorted-arrays/#/description
 * Created by aghoneim on 4/18/17.
 */
public class MedianTwoSortedArrays {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        if (nums1 == null) {
            nums1 = new int[]{};
        }

        if (nums2 == null) {
            nums2 = new int[]{};
        }

        int size = nums1.length + nums2.length;
        if (size > 0) {
            int firstMedian = 0 , secondMedian = 0;
            int i = 0, j = 0, medianIndex = 0;
            int middle = (size - 1)/ 2;
            while (i < nums1.length || j < nums2.length) {
                int current = 0;
                if (i >= nums1.length) {
                    current = nums2[j];
                    j++;
                } else if (j >= nums2.length) {
                    current = nums1[i];
                    i++;
                } else if (nums1[i] < nums2[j]) {
                    current = nums1[i];
                    i++;
                } else {
                    current = nums2[j];
                    j++;
                }
                if (medianIndex == middle) {
                    firstMedian = current;
                } else if (medianIndex == middle + 1) {
                    secondMedian = current;
                } else if(medianIndex > middle) {
                    break;
                }
                medianIndex++;
            }
            return size % 2 == 0 ? (firstMedian + secondMedian) / 2.0 : firstMedian;
        } else {
            return 0;
        }
    }


    public static void main(String [] args) {
        MedianTwoSortedArrays m = new MedianTwoSortedArrays();
        System.out.println(m.findMedianSortedArrays(new int[]{0, 2, 3, 3}, new int[]{ 4, 5}));
    }

}

File: src/main/kotlin/arrays/MedianTwoSortedArrays.kt
package arrays

/**
 * https://leetcode.com/problems/median-of-two-sorted-arrays/#/description
 * Created by aghoneim on 4/18/17.
 */
class MedianTwoSortedArrays {
    fun findMedianSortedArrays(nums1: IntArray?, nums2: IntArray?): Double {
        var nums1 = nums1
        var nums2 = nums2
        if (nums1 == null) {
            nums1 = intArrayOf()
        }
        if (nums2 == null) {
            nums2 = intArrayOf()
        }
        val size = nums1.size + nums2.size
        return if (size > 0) {
            var firstMedian = 0
            var secondMedian = 0
            var i = 0
            var j = 0
            var medianIndex = 0
            val middle = (size - 1) / 2
            while (i < nums1.size || j < nums2.size) {
                var current = 0
                if (i >= nums1.size) {
                    current = nums2[j]
                    j++
                } else if (j >= nums2.size) {
                    current = nums1[i]
                    i++
                } else if (nums1[i] < nums2[j]) {
                    current = nums1[i]
                    i++
                } else {
                    current = nums2[j]
                    j++
                }
                if (medianIndex == middle) {
                    firstMedian = current
                } else if (medianIndex == middle + 1) {
                    secondMedian = current
                } else if (medianIndex > middle) {
                    break
                }
                medianIndex++
            }
            if (size % 2 == 0) (firstMedian + secondMedian) / 2.0 else firstMedian.toDouble()
        } else {
            0.0
        }
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val m = MedianTwoSortedArrays()
            println(m.findMedianSortedArrays(intArrayOf(0, 2, 3, 3), intArrayOf(4, 5)))
        }
    }
}